source/common/upstream/eds.cc

```diff
@@ -61,6 +61,15 @@ void EdsClusterImpl::onConfigUpdate(const ResourceVector& resources, const std::
     throw EnvoyException(fmt::format("Unexpected EDS cluster (expecting {}): {}", cluster_name_,
                                      cluster_load_assignment.cluster_name()));
   }
+
+  HostVector all_current_hosts;
```

Comment:
Although I agree with the intention here, I'm concerned about the perf impact of this change, given that we are already having perf issues with this code which is starting to be addressed in https://github.com/envoyproxy/envoy/pull/3941 by @rgs1. 

I think we might need to holistically step back and try to figure out the right path forward for solving this issue while also keeping/improving perf. Concretely, I wonder if we should be keeping an "all hosts" unordered_map keyed by address in parallel to the per-priority vectors? This would allow us to quickly do checks for whether a host already exists in a different priority. 

I will also review https://github.com/envoyproxy/envoy/pull/3941 today and will probably have additional thoughts after I do that so will circle back then.

Reply:
Yeah I thought about the map approach but I wasn't sure about the perf requirements of this code path so I wanted to float the naive solution first.

Thinking out loud a bit: if we kept such a map we could potentially simplify/speed up this code quite a bit by making addresses unique across priorities: instead of iterating over the current hosts we'd just check the map for an existing host with that address. This would make moving a host between priorities no longer require additional code and would eliminate the O(n) scan over existing addresses. A nice thing here is that we could reuse the `Host` from another priority when a host moves, saving us some heap allocations.

Alternatively if we want to maintain the current behavior of allowing duplicate addresses across priorities: we could maintain two maps `std::unordered_map<AddressPtr, Host>` and `std::unordered_map<std::pair<AddressPtr, Integer>, Host>` where one would let us look up hosts from any priority (for the "all hosts" check) and the other would let us look up hosts within the same priority. Essentially this transforms the code in my PR to two map lookups instead of two array scans per priority. 

Unclear to me whether the host uniqueness only within priority invariant is intentional or just a result of the implementation, so not sure how important it is to maintain.

Reply:
> Unclear to me whether the host uniqueness only within priority invariant is intentional or just a result of the implementation, so not sure how important it is to maintain.

I actually don't know if this was intentional or not. I feel like it was but I'm not sure. IMO it would be great to simplify it if we can. @mpuncel @htuch @alyssawilk @zuercher do you remember the history here? I don't.

Reply:
I don't have historical context on the code, but that plan sounds reasonable. I can't think of a reason you'd want multiple endpoints with the same address in the same cluster since there is proper support for load balancing weight.